README.md

The PenTesters Framework (PTF)

A TrustedSec Project - Copyright 2016

Written by: David Kennedy (@HackingDave)

https://www.trustedsec.com

Twitter: @TrustedSec, @HackingDave

The PenTesters Framework (PTF) is a Python script designed for Debian/Ubuntu/ArchLinux based distributions to create a similar and familiar distribution for Penetration Testing. As pentesters, we've been accustom to the /pentest/ directories or our own toolsets that we want to keep up-to-date all of the time. We have those "go to" tools that we use on a regular basis, and using the latest and greatest is important.

PTF attempts to install all of your penetration testing tools (latest and greatest), compile them, build them, and make it so that you can install/update your distribution on any machine. Everything is organized in a fashion that is cohesive to the Penetration Testing Execution Standard (PTES) and eliminates a lot of things that are hardly used. PTF simplifies installation and packaging and creates an entire pentest framework for you. Since this is a framework, you can configure and add as you see fit. We commonly see internally developed repos that you can use as well as part of this framework. It's all up to you.

The ultimate goal is for community support on this project. We want new tools added to the github repository. Submit your modules. It's super simple to configure and add them and only takes a few minute.

#Instructions:

First check out the config/ptf.config file which contains the base location of where to install everything. By default this will install in the /pentest directory. Once you have that configured, move to running PTF by typing ./ptf (or python ptf).

This will put you in a Metasploitesque type shell which has a similar look and feel for consistency. Show modules, use <modules>, etc. are all accepted commands. First things first, always type help or ? to see a full list of commands.

For a video tutorial on how to use PTF, check out our Vimeo page here: https://vimeo.com/137133837

###Update EVERYTHING!

If you want to install and/or update everything, simply do the following:

./ptf

use modules/install_update_all

yes

This will install all of the tools inside of PTF. If they are already installed, this will iterate through and update everything for you automatically.

You can also show options to change information about the modules.

If you only want to install only for example exploitation tools, you can run:

./ptf
use modules/exploitation/install_update_all

This will only install the exploitation modules. You can do this for any module category.

#Modules:

First, head over to the modules/ directory, inside of there are sub directories based on the Penetration Testing Execution Standard (PTES) phases. Go into those phases and look at the different modules. As soon as you add a new one, for example testing.py, it will automatically be imported next time you launch PTF. There are a few key components when looking at a module that must be completed.

Below is a sample module

AUTHOR="David Kennedy (ReL1K)"

DESCRIPTION="This module will install/update the Browser Exploitation Framework (BeEF)"

INSTALL_TYPE="GIT"

REPOSITORY_LOCATION="https://github.com/beefproject/beef"

X64_LOCATION="https://github.com/something_thats_x64_instead_of_x86

INSTALL_LOCATION="beef"

DEBIAN="ruby1.9.3,sqlite3,ruby-sqlite3"

ARCHLINUX = "arch-module,etc"

BYPASS_UPDATE="NO"

AFTER_COMMANDS="cd {INSTALL_LOCATION},ruby install-beef"

LAUNCHER="beef"

###Module Development:

All of the fields are pretty easy, on the repository locations, you can use GIT, SVN or FILE. Fill in the depends, and where you want the install location to be. PTF will take where the python file is located (for example exploitation) and move it to what you specify in the PTF config (located under config). By default it installs all your tools to /pentest/PTES_PHASE/TOOL_FOLDER

Note in modules, you can specify after commands {INSTALL_LOCATION}. This will append where you want the install location to go when using after commands.

You also have the ability for repository locations to specify both a 32 bit and 64 bit location. Repository location should always be the x86 download path. To add a 64 bit path for a tool, specify X64_LOCATION and give it a URL. When PTF launches it will automatically detect the architecture and attempt to use the x64 link instead of the x86.

Note that ArchLinux packages are also supported, it needs to be specified for both DEBIAN and ARCH in order for it to be properly installed on either platform in the module

###BYPASS UPDATES:

When using traditional git or svn as a main method, what will happen after a module is installed is it will just go and grab the latest version of the tool. With after commands, normally when installing, you may need to run the after commands after each time you update. If you specify bypass updates to YES (BYPASS_UPDATE="YES"), each time the tool is run, it will check out the latest version and still run after commands. If this is marked to no, it will only git pull the latest version of the system. For "FILE" options, it is recommended to always use BYPASS_UPDATE="YES" so that it will overwrite the files each time.

###After Commands:

After commands are commands that you can insert after an installation. This could be switching to a directory and kicking off additional commands to finish the installation. For example in the BEEF scenario, you need to run ruby install-beef afterwards.  Below is an example of after commands using the {INSTALL_LOCATION} flag.
 
AFTER_COMMANDS="cp config/dict/rockyou.txt {INSTALL_LOCATION}"

For AFTER_COMMANDS that do self install (don't need user interaction).

###Automatic Launchers

The flag LAUNCHER= in modules is optional. If you add LAUNCHER="setoolkit" for example, PTF will automatically create a launcher for the tool under /usr/local/bin/. In the setoolkit example, when run - PTF will automatically create a file under /usr/local/bin/setoolkit so you can launch SET from anywhere by simply typing setoolkit. All files will still be installed under the appropriate categories, for example /pentest/exploitation/setoolkit however an automatic launcher will be created.

You can have multiple launchers for an application - for example Metasploit you may want msfconsole, msfvenom, etc. etc. In order to add multiple ones, simply put a "," between them. For example LAUNCHER="msfconsole,msfvenom". This would create launchers for both.

### Automatic Command Line

You can also just run ./ptf --update-all and it will automatically update everything for you without having to do into the framework

### IGNORE Modules or Categories

The "IGNORE_THESE_MODULES=" config option can be found under config/ptf.config in the PTF root directory. This will ignore modules and not install them - everything is comma separated and based on name - example: modules/exploitation/metasploit,modules/exploitation/set or entire module categories, like modules/code-audit/*,/modules/reporting/*


CHANGELOG
~~~~~~~~~~~~~~~~~
version 1.7
~~~~~~~~~~~~~~~~~

* added samba-client as prereq for ridenum

~~~~~~~~~~~~~~~~~
version 1.6.4
~~~~~~~~~~~~~~~~~

* added multiple new password recovery tools thanks to Martin Bos (@CantComputer)

~~~~~~~~~~~~~~~~~
version 1.6.3
~~~~~~~~~~~~~~~~~

* fixed install location to properly work
* removed sslsplit, python2-dpkt, python-pypcap
* added new function for detection in Kali
* added smart detection for Kali to not install Metasploit on there - packages are different
* added pep8 compliance to PTF
* added compatibility to python3 (doesn't mean modules will support python3 - just that ptf is python 3 if you wish

~~~~~~~~~~~~~~~~~
version 1.6.2
~~~~~~~~~~~~~~~~~

* fixed an issue with hydra not installing ldap-utils
* ssl certificate issue fixed with openvas (cert does not match)

~~~~~~~~~~~~~~~~~
version 1.6.1
~~~~~~~~~~~~~~~~~

* add a remove from Gemfile.lock in order to ensure proper git pulls

~~~~~~~~~~~~~~~~~
version 1.6
~~~~~~~~~~~~~~~~~

* move nmap-dev to nmap
* fixed spraywmi depends for Ubuntu
* moved ridenum from exploitation to intelligence-gathering
* fixed an issue that would not import all of the arguments when using launchers
* changed git to svn language to fix typo
* added Python3 compatibility and support
* added check for github, if new update occurs, run through after commands
* added autocheck for svn if new updates occur, run through after commands
* removed build-dep as a non-standard module in sslscan
* added better handling for all updates - after commands will be now run anytime theres a module update either via SVN or GIT - this already happens for FILE. If something is already up-to-date it won't run after_commands
* removed prompt for empire
* added build gem bundles for smbexec but its still a mess - installer is broke, requires a million external depends - will need to work on this one later
* just completely removed smbexec - hasn't been updated in over a year - no support and super broke

~~~~~~~~~~~~~~~~~
version 1.5.2
~~~~~~~~~~~~~~~~~

* adds ignore module or entire categories - pull request from vdbaan (awesome work!)

~~~~~~~~~~~~~~~~~
version 1.5.1
~~~~~~~~~~~~~~~~~

* added newline after each module install
* fixed automatic apt-get when not on debian
* fixed an undefined message when module would not have a OStype

~~~~~~~~~~~~~~~~~
version 1.5
~~~~~~~~~~~~~~~~~

* added kismet wireless
* added dnscat2 (thanks ryd)
* fix for discover (thanks jayw0k)
* fix dnsrecon (thanks jayw0k)
* fix theharvester (thanks jayw0k)
* pep8 compliance (thanks vdbaan)
* fix for auto complete (thanks vdbaan)
* udp proto added (thanks vdbaan)
* what web added (thanks vdbaan)
* added install_update_all to each individual module section (thanks vdbaan)
* added simplyemail - recon/mail harvester
* added pip - mainframe exploitation

~~~~~~~~~~~~~~~~~
version 1.4.2
~~~~~~~~~~~~~~~~~

* added new IGNORE_THESE_MODULES config option to ignore modules you dont want to install (awesome pull request from vdbaan)
* added fierce module
* added config option for ignore these modules
* added updatedb to automatic update functionality
* added onesixtyone (vdbaan)
* added yapscan (vdbaan)
* added arachni (vdbaan)
* added ./ptf --update-all which will automatically update without going in the framework

~~~~~~~~~~~~~~~~~
version 1.4.1
~~~~~~~~~~~~~~~~~

* added hydra (thanks jayw0k)
* added enum4linux (vdbaan)
* added testssl (vdbaan)
* fixed autoconf error on install with ftpmap

~~~~~~~~~~~~~~~~~
version 1.4
~~~~~~~~~~~~~~~~~

* updated metasploit installer to use ppa instead of rvm - fixes a number of issues on Ubuntu

~~~~~~~~~~~~~~~~~
version 1.3
~~~~~~~~~~~~~~~~~

* spent a lot of time fixing the Metasploit install - looks to be bullet proof now and installs properly
* binwalk added and fix for control-d EOF (thanks Jens!)
* added spraywmi to tools

~~~~~~~~~~~~~~~~~
version 1.2.3
~~~~~~~~~~~~~~~~~

* number of additions to modules to include fedora compatibility (awesome work fnoris!)

~~~~~~~~~~~~~~~~~
version 1.2.2
~~~~~~~~~~~~~~~~~

* fixed print statement when using x64 base
* fixed invalid path error when using after commands
* fixed hcon to install properly
* fixed dradisframework by adding libsqlite-dev (thanks Kos)
* fixed wine install by adding i386 packages (thanks Kos)
* fixed prompt to not be used when using empire (thanks Kos)

~~~~~~~~~~~~~~~~~
version 1.2.1
~~~~~~~~~~~~~~~~~

* fixed an issue where empty directory would throw an error
* fixed the code-audit/splint module - was broke and not properly made

~~~~~~~~~~~~~~~~~
version 1.2
~~~~~~~~~~~~~~~~~

* added dnsenum to modules to intelligence gathering
* added windows-exploit-suggester to modules intelligence gathering
* added launcher for recon-ng
* added jboss-autopwn (thanks firebitsbr)
* added initial commit for openBSD - not fully implemented yet (thanks firebitsbr)
* added full commit for OpenBSD - awesome work firebitsbr
* added if it isn't a specific path - it'll use normal launchers

~~~~~~~~~~~~~~~~~
version 1.1
~~~~~~~~~~~~~~~~~

* added new module option called X64_LOCATION which will pull a x64 bit image if they are available on the tool, updated documentation to reflect changes
* removed unused print on tool when creating launcher
* added support for Arch thanks to pull request from Hypsurus

~~~~~~~~~~~~~~~~~
version 1.0.5
~~~~~~~~~~~~~~~~~

* fix for nmap (git pull)
* add for medusa (git pull)

~~~~~~~~~~~~~~~~~
version 1.0.4
~~~~~~~~~~~~~~~~~

* updated launchers to handle command line arguments (./<tool> $*) - thanks to tdelagrange
* added automatic overwrite of launchers in order to force updates to launchers when needed

~~~~~~~~~~~~~~~~~
version 1.0.3
~~~~~~~~~~~~~~~~~

* added ignore file for git (thanks KryptonSecurity)
* new folder for code audit (thanks KryptonSecurity)
* new module added java decompiler - CFR (Thanks KryptonSecurity)

src/core.py
#!/usr/bin/env python
################################
# Core functions for PTF
################################
import os
import subprocess
import select
import readline
import glob
import platform
import urllib
import sys

# tab completion
def complete(text, state):
    a = (glob.glob(text + '*') + [None])[state].replace("__init__.py", "").replace(".py", "").replace("LICENSE", "").replace(
        "README.md", "").replace("config", "").replace("ptf", "").replace("readme", "").replace("src", "").replace("         ", "") + "/"
    a = a.replace("//", "/")
    if os.path.isfile(a[:-1] + ".py"):
        return a[:-1]
    else:
        return a

readline.set_completer_delims(' \t\n;')
readline.parse_and_bind("tab: complete")
readline.set_completer(complete)
# end tab completion

# color scheme for core


class bcolors:
    PURPLE = '\033[95m'
    CYAN = '\033[96m'
    DARKCYAN = '\033[36m'
    BLUE = '\033[94m'
    GREEN = '\033[92m'
    YELLOW = '\033[93m'
    RED = '\033[91m'
    BOLD = '\033[1m'
    UNDERL = '\033[4m'
    ENDC = '\033[0m'
    backBlack = '\033[40m'
    backRed = '\033[41m'
    backGreen = '\033[42m'
    backYellow = '\033[43m'
    backBlue = '\033[44m'
    backMagenta = '\033[45m'
    backCyan = '\033[46m'
    backWhite = '\033[47m'

# custom parser for zaproxy
def zaproxy():
    file = urllib.urlopen('https://raw.githubusercontent.com/zaproxy/zap-admin/master/ZapVersions.xml')
    data = file.readlines()
    file.close()
    for url in data:
        if "Linux.tar.gz" in url and "<url>" in url: return url.rstrip().replace("<url>", "").replace("</url>", "").strip()


# get the main SET path
def definepath():
    if os.path.isfile("ptf"):
        return os.getcwd()

    else:
        if os.path.isdir("/usr/share/ptf/"):
            return "/usr/share/ptf/"
        else:
            return os.getcwd()

# main status calls for print functions


def print_status(message):
    print((bcolors.GREEN) + (bcolors.BOLD) + \
        ("[*] ") + (bcolors.ENDC) + (str(message)))


def print_info(message):
    print((bcolors.BLUE) + (bcolors.BOLD) + \
        ("[-] ") + (bcolors.ENDC) + (str(message)))


def print_info_spaces(message):
    print((bcolors.BLUE) + (bcolors.BOLD) + \
        ("  [-] ") + (bcolors.ENDC) + (str(message)))


def print_warning(message):
    print((bcolors.YELLOW) + (bcolors.BOLD) + \
        ("[!] ") + (bcolors.ENDC) + (str(message)))


def print_error(message):
    print((bcolors.RED) + (bcolors.BOLD) + \
        ("[!] ") + (bcolors.ENDC) + (bcolors.RED) + \
        (str(message)) + (bcolors.ENDC))

def set_title(title):
	sys.stdout.write("\x1b]2;%s\x07" % title)

# count all of the modules


def count_modules():
    modules_path = definepath() + "/modules/"
    counter = 0
    for path, subdirs, files in os.walk(modules_path):
        for name in files:
            filename = os.path.join(path, name)
            if not "__init__.py" in filename:
                counter = counter + 1
    return counter

# version information
grab_version = "1.8"

# banner
banner = bcolors.RED + r"""

                     ______  __ __    ___
                    |      T|  T  T  /  _]
                    |      ||  l  | /  [_
                    l_j  l_j|  _  |Y    _]
                      |  |  |  |  ||   [_
                      |  |  |  |  ||     T
                      l__j  l__j__jl_____j

 ____     ___  ____   ______    ___   _____ ______    ___  ____    _____
|    \   /  _]|    \ |      T  /  _] / ___/|      T  /  _]|    \  / ___/
|  o  ) /  [_ |  _  Y|      | /  [_ (   \_ |      | /  [_ |  D  )(   \_
|   _/ Y    _]|  |  |l_j  l_jY    _] \__  Tl_j  l_jY    _]|    /  \__  T
|  |   |   [_ |  |  |  |  |  |   [_  /  \ |  |  |  |   [_ |    \  /  \ |
|  |   |     T|  |  |  |  |  |     T \    |  |  |  |     T|  .  Y \    |
l__j   l_____jl__j__j  l__j  l_____j  \___j  l__j  l_____jl__j\_j  \___j

 _____  ____    ____  ___ ___    ___  __    __   ___   ____   __  _
|     ||    \  /    T|   T   T  /  _]|  T__T  T /   \ |    \ |  l/ ]
|   __j|  D  )Y  o  || _   _ | /  [_ |  |  |  |Y     Y|  D  )|  ' /
|  l_  |    / |     ||  \_/  |Y    _]|  |  |  ||  O  ||    / |    \
|   _] |    \ |  _  ||   |   ||   [_ l  `  '  !|     ||    \ |     Y
|  T   |  .  Y|  |  ||   |   ||     T \      / l     !|  .  Y|  .  |
l__j   l__j\_jl__j__jl___j___jl_____j  \_/\_/   \___/ l__j\_jl__j\_j
"""

banner += bcolors.ENDC + """
		     The"""
banner += bcolors.BOLD + """ PenTesters """
banner += bcolors.ENDC + """Framework\n\n"""

banner += """        		  """ + bcolors.backBlue + \
    """Version: %s""" % (grab_version) + bcolors.ENDC + "\n"

banner += bcolors.YELLOW + bcolors.BOLD + """		      Codename: """ + \
    bcolors.BLUE + """Tool Depot""" + "\n"

banner += """		       """ + bcolors.ENDC + bcolors.backRed + \
    """Red Team Approved""" + bcolors.ENDC + "\n"

banner += """        	    A project by """ + bcolors.GREEN + bcolors.BOLD + \
    """Trusted""" + bcolors.ENDC + bcolors.BOLD + """Sec""" + bcolors.ENDC + "\n"

banner += """		 Written by: """ + bcolors.BOLD + \
    """Dave Kennedy (ReL1K)""" + bcolors.ENDC + "\n"
banner += """		Twitter: """ + bcolors.BOLD + \
    """@HackingDave, @TrustedSec""" + bcolors.ENDC + "\n"
banner += """                    """ + bcolors.BOLD + """https://www.trustedsec.com
        """ + bcolors.ENDC
banner += bcolors.BOLD + """\n              The easy way to get the new and shiny.
""" + bcolors.ENDC + "\n"
banner += "             Total module/tool count within PTF: " + \
    bcolors.BOLD + str(count_modules()) + bcolors.ENDC + "\n"
banner += """
All tools are downloaded directly from the developers websites as-is.  PTF 
doesn't perform any type of source code analysis or verification on the tools.
You should run these after performing your own analysis of the tools and ensure 
you trust the parties. PTF only adds tools that are well-known typically in the 
security industry but that does not negate the risk. This is no different than 
any other tool distribution platform, operating system, or anything you would 
download from the Internet.\n"""

# check the config file and return value


def check_config(param):
    fileopen = open("%s/config/ptf.config" % (definepath()), "r")
    for line in fileopen:
        # if the line starts with the param we want then we are set, otherwise
        # if it starts with a # then ignore
        if line.startswith(param) != "#":
            if line.startswith(param):
                line = line.rstrip()
                # remove any quotes or single quotes
                line = line.replace('"', "")
                line = line.replace("'", "")
                line = line.split("=")
                return line[1]

# parser module for module and term
def module_parser(filename, term):
    # if the file exists
    if os.path.isfile(filename) and not "install_update_all" in filename:

        # set a base counter
        counter = 0

        # open the file
        fileopen = open(filename, "r")
        # iterate through the file
        for line in fileopen:
            # strip any bogus stuff
            line = line.rstrip()
            # if the line starts with the term
            if line.startswith(term):
                line = line.replace(term + "=", "")
                line = line.replace('"', "", 2)
                # reflect we hit this and our search term was found
                counter = 1
                return line

        # if there was no search term identified for the module
        if counter == 0:
            filename_short = filename.replace(definepath() + "/", "")
            filename_short = filename_short.replace(".py", "")
            if term != "BYPASS_UPDATE":
                if term != "LAUNCHER":
                    if filename_short != "install_update_all":
                        if term != "X64_LOCATION":
                              if not "__init__" in filename_short:
                                    print_error("Warning, module %s was found but contains no %s field." % (filename_short, term))
                                    print_error("Check the module again for errors and try again.")
                                    print_error("Module has been removed from the list.")

            return ""

    # if the file isn't there
    if not os.path.isfile(filename):
        return None

# help menu for PTF


def show_help_menu():
    print(("Available from main prompt: " + bcolors.BOLD + "show modules" + bcolors.ENDC + "," + bcolors.BOLD + " show <module>" +
           bcolors.ENDC + "," + bcolors.BOLD + " search <name>" + bcolors.ENDC + "," + bcolors.BOLD + " use <module>" + bcolors.ENDC))
    print(("Inside modules:" + bcolors.BOLD + " show options" + bcolors.ENDC + "," +
           bcolors.BOLD + " set <option>" + bcolors.ENDC + "," + bcolors.BOLD + "run" + bcolors.ENDC))
    print(("Additional commands: " + bcolors.BOLD + "back" + bcolors.ENDC + "," + bcolors.BOLD + " help" + bcolors.ENDC + "," +
           bcolors.BOLD + " ?" + bcolors.ENDC + "," + bcolors.BOLD + " exit" + bcolors.ENDC + "," + bcolors.BOLD + " quit" + bcolors.ENDC))
    print(("Update or Install: " + bcolors.BOLD + "update" + bcolors.ENDC + "," + bcolors.BOLD + " upgrade" +
           bcolors.ENDC + "," + bcolors.BOLD + " install" + bcolors.ENDC + "," + bcolors.BOLD + " run" + bcolors.ENDC))

# exit message for PTF


def exit_ptf():
    print_status("Exiting PTF - the easy pentest platform creation framework.")
    set_title("Hack the Planet!")


# this is the main handler to check what distribution we are using
# if it can't find it, will default to manual install base
def profile_os():
    # if we are running a debian variant
    if os.path.isfile("/usr/bin/apt-get"):
        return "DEBIAN"
    if os.path.isfile("/etc/arch-release"):
        return "ARCHLINUX"
    if os.path.isfile("/etc/fedora-release"):
        return "FEDORA"
    # will add support for more operating systems later

    # else use custom
    else:
        return "CUSTOM"


# standard log write out
def logging(log):
    # grab the log path
    logpath = check_config("LOG_PATH=")
    # if the file isn't there, create it
    if not os.path.isfile(logpath):
        filewrite = open(logpath, "w")
        filewrite.write("")
        filewrite.close()
    # open for append
    filewrite = open(logpath, "a")
    # write it out
    filewrite.write(log)
    # close the file
    filewrite.close()

# this will install all the proper locations for


def prep_install():
    if not os.path.isfile(os.getenv("HOME") + "/.ptf"):
        print_status("This appears to be your first time using PTF.")
        print_status("Creating output directory to: " +
                     os.getenv("HOME") + "/.ptf")
        os.makedirs(os.getenv("HOME") + "/.ptf")


def home_directory():
    return os.getenv("HOME") + "/.ptf"

# this will run commands after an install or update on a module


def after_commands(filename, install_location):
    from src.commands import after_commands
    commands = module_parser(filename, "AFTER_COMMANDS")
    if commands != "":
        # here we check if install location needs to be added
        if "{INSTALL_LOCATION}" in commands:
            commands = commands.replace("{INSTALL_LOCATION}", install_location)
        print_status(
            "Running after commands for post installation requirements.")
        after_commands(commands, install_location)
        print_status("Completed running after commands routine..")

# launcher - create launcher under /usr/local/bin


def launcher(filename, install_location):
    launcher = module_parser(filename, "LAUNCHER")

    # if its optional
    if launcher == None:
        launcher = ""
    if launcher != "":
        # create a launcher if it doesn't exist
        base_launcher = 0
        if "," in launcher:
            launcher = launcher.split(",")
        for launchers in launcher:

            # means there was just one launcher
            if len(launchers) == 1:
                launchers = launcher
                base_launcher = 1

            if os.path.isfile("/usr/local/bin/" + launchers):
                os.remove("/usr/local/bin/" + launchers)
            if not os.path.isfile("/usr/local/bin/" + launchers):

                # base launcher filename
                point = ""

                # make sure the actual launcher is there with known filetypes
                if os.path.isfile(install_location + "/" + launchers):
                    # specific launcher file
                    point = "./" + launchers
                    file_point = launchers

                # check for Python
                if os.path.isfile(install_location + "/" + launchers + ".py"):
                    point = "./" + launchers + ".py"
                    file_point = launchers + ".py"

                # check for Ruby
                if os.path.isfile(install_location + "/" + launchers + ".rb"):
                    point = "./" + launchers + ".rb"
                    file_point = launchers + ".rb"

                # check for Perl - ew Perl. Ew ew ew ew ew ew =)
                if os.path.isfile(install_location + "/" + launchers + ".pl"):
                    point = "./" + launchers + ".pl"
                    file_point = launchers + ".pl"

                # check for bash
                if os.path.isfile(install_location + "/" + launchers + ".sh"):
                    point = "./" + launchers + ".sh"
                    file_point = launchers + ".sh"

                # check of executable, then flag wine
                if os.path.isfile(install_location + "/" + launchers + ".exe"):
                    point = "wine cmd /c start " + launchers + ".exe"
                    file_point = launchers + ".exe"

                # normal launcher
                if os.path.isfile(install_location + "/" + launchers):
                    point = "./" + launchers
                    file_point = launchers

                # if we found filetype
                if point != "":
                    filewrite = open("/usr/local/bin/" + launchers, "w")
                    filewrite.write('#!/bin/sh\ncd %s\nchmod +x %s\n%s $*\n' %
                                    (install_location, file_point, point))
                    filewrite.close()
                    subprocess.Popen("chmod +x /usr/local/bin/%s" %
                                     (launchers), shell=True).wait()
                    print_status(
                        "Created automatic launcher, you can run the tool from anywhere by typing: " + launchers)

            # just need to do this once
            if base_launcher == 1:
                break

# search functionality here


def search(term):
    term = term.replace("search ", "")
    module_files = []
    if "update" in term or "install" in term:
        module_files.append("modules/install_update_all")

    else:
        for dirpath, subdirs, files in os.walk("modules/"):
            for x in files:
                if x.endswith(".py"):
                    if not "__init__.py" in x:
                        path = os.path.join(dirpath, x)
                        if term in path:
                            x = x.replace(".py", "")
                            module_files.append(os.path.join(dirpath, x))

                        if not term in path:
                            data = open(path, "r").readlines()
                            # normally just searched entire file, but we don't
                            # want to search # lines
                            for line in data:
                                line = line.rstrip()
                                if term in line:
                                    if not line.startswith("#"):
                                        x = x.replace(".py", "")
                                        module_files.append(
                                            os.path.join(dirpath, x))
                                        break
    if module_files != []:
        print_status("Search results below:")
        for modules in module_files:
            print (modules)

    else:
        print_warning("Search found no results.")


# auto update packages
def auto_update():
    # if we want to do auto update
    check = check_config("AUTO_UPDATE=").lower()
    if check == "on":
        print_status(
            "Auto updating is turned to on, this will install normal package updates for you...")
        print_status(
            "If you want to turn this off, go to the PTF directory and go to config and change AUTO_UPDATE")
        if profile_os() == "DEBIAN":
            subprocess.Popen(
                "sudo apt-get update && sudo apt-get -y upgrade && sudo apt-get dist-upgrade -y && sudo apt-get autoremove -y && apt-get autoclean -y && updatedb", shell=True).wait()
        print_status(
            "Finished with normal package updates, moving on to the tools section..")
    else:
        print_status(
            "Auto updating for packages is turned off, to enable go to PTF and config directory and turn AUTO_UPDATE to ON.")

# check if a blank directory exists


def check_blank_dir(path):

    if os.path.isdir(path):
        if os.listdir(path) == []:
            print_status(
                "Detected an empty folder, purging and re-checking out...")
            subprocess.Popen("rm -rf %s" % (path), shell=True).wait()

        # we put a second one in there in case the path was removed from above
        if os.path.isdir(path):
            if os.listdir(path) == ['.git', '.gitignore']:
                print_status(
                    "Detected an empty folder, purging and re-checking out...")
                subprocess.Popen("rm -rf %s" % (path), shell=True).wait()

# do platform detection on 32 or 64 bit


def arch():
    return str(platform.architecture()[0])

# check to see if we are running kali linux


def check_kali():
    if os.path.isfile("/etc/apt/sources.list"):
        kali = open("/etc/apt/sources.list", "r")
        kalidata = kali.read()
        if "kali" in kalidata:
            return "Kali"
        # if we aren't running kali
        else:
            return "Non-Kali"
    else:
        print("[!] Not running a Debian variant..")
        return "Non-Kali"
